Matching DEIMOS slits to slit-mask design IDs

deborape opened the pull request:
First step to ingest mask design information.

-- edits to DEIMOS optical model
-- added "amap" and "bmap" files in pypeit/data/static_calibs/keck_deimos/
-- added `maskdesign_matching` to `EdgeTraceSet.auto_trace()`. It is switched ON only if the instrument is `keck_deimos`
-- added pypeit/core/slitdesign_matching.py, which is a collection of script that perform the edge matching
-- recorded `maskdef_id` in `SlitTraceSet`

I am running the tests now.

PS: I have noticed that  pypeit/spectrographs/opticalmodel.py has a lot of edits because of an automatic re-formatting of the text (pyCharm re-adjusted the spaces). Sorry about that.

deborape commented:
`./pypeit_test develop -i keck_deimos` is failing for some datasets. 
Also Travis is complaining for `test_maskdesign_matching.py`. 
I am working on those!

profxj reviewed (changes requested):
Holy cow!  Sorry that your first PR for PypeIt needed to be so long.
But the work is greatly appreciated.

I've put in some Python suggestions here in there, especially in
places where I thought it would turn the IDL -> Python conversion
into something a bit more Pythonic.

I think the only logical issue I raised is what to do when there are 
duplicated `ind` found in `slit_match()`.  If I read the code correctly,
the first one is assumed correct and the rest are trimmed.  One might
argue they should all be trimmed, but maybe I am not quite following
the logic.  :)

And feel free to punt items that I suggested to a later PR.
Thanks!

deborape commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,226 @@
+    """
+    nbest = int(x_det.shape[0] * .85)
+    if xlag_range is not None:
+        xlag = numpy.arange((xlag_range[1]-xlag_range[0])/step+1)*step+xlag_range[0]
```
Done. Thanks!

deborape commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,226 @@
+    sdev = numpy.zeros((xlag.shape[0]))
+
+    # -------- Loop over lags
+    for j in range(xlag.shape[0]):
```
Done

deborape commented:
Unit tests passed
`================= 246 passed, 662 warnings in 633.70s (0:10:33) =====================`

deborape commented on pypeit/edgetrace.py:
```diff
@@ -3970,326 +3988,235 @@ def fully_masked_traces(self, flag=None, exclude=None):
+        wh = model_xt != model_xb
+        switched = np.mean(model_xt[wh] - model_xb[wh]) < 0
+        # Matching
+        if switched == False:
```
Done.

deborape commented:
I made some small fixes and now `./pypeit_test develop -i keck_deimos` also passes.
`--- PYPEIT DEVELOPMENT SUITE PASSED 7/7 TESTS (Masters ignored) ---`

I will slowly work on the other comments.

jhennawi commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,229 @@
+        if x_model[wkeep].size<2:
+            print('Working between {} and {}'.format(min_x_det+xlag_range[0], max_x_det+xlag_range[1]))
+            print('Not enough lines to run!!!')
+            sdev=1.E10
```
Change to 1e10 to be more consistent with pep style.

jhennawi commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,229 @@
+    residual = (x_det-x_model[ind]) - numpy.median(x_det-x_model[ind])
+    weights = numpy.zeros(residual.size, dtype=int)
+    weights[numpy.abs(residual) < 100.] = 1
+    if weights.sum()==0: weights=numpy.ones(residual.size, dtype=int)
```
Indent here in if statement to be consistent with PEP python coding style.

jhennawi commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,226 @@
+    if weights.sum()==0: weights=numpy.ones(residual.shape[0], dtype=int)
+    coeff, cov= curve_fit(linefit, x_model[ind]*weights, x_det, p0=[0,1])
+    coeff=numpy.append(coeff, 0)
+    yfit = slit_coeff_eval(x_model[ind], x_model[ind]*0., coeff)
```
Can we use np.polyfit here instead of the curve_fit. This is the more pythonic way to fit a straight line. Also, if you need to reject outliers, we should consider using robust_fit.py

deborape replied:
Made some changes here to use instead `fitting.robust_fit`. I also removed `slit_coeff_eval`

jhennawi replied:
I still see the curve fit call above?

jhennawi commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,229 @@
+
+
+
+def plot_matches(edgetrace, ind, x_model, x_det, yref, slit_index, edge=None):
```
Docs here and everywhere.

jhennawi commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,229 @@
+                                     fontsize=8, horizontalalignment='center')
+    plt.xlabel('Spatial pixels')
+    plt.ylabel('Spectral pixels')
+    plt.xlim(0, 2218)
```
These appear to be hard coded to DEIMOS. Grab the sizes of the arrays you are using in this code and decide based on them how to scale the axes.

deborape replied:
Done.

jhennawi commented on pypeit/spectrographs/keck_deimos.py:
```diff
@@ -703,24 +708,64 @@ def _grating_orientation(slider, ruling, tilt):
+
+        mp_dir = resource_filename('pypeit', 'data/static_calibs/keck_deimos/')
+
+        if slider in [3,4]:
```
IDL save files are in general not a good long term solution for storing data. I realize that there exists a method to read them in, but we should be reading in these maps from the save files, and then writing them out to fits or another stable binary format, and then reading those in in PypeIt. My suggestion is that you write a utility routine in core that does this and then push the converted fits files to the pypeit data repository.

deborape replied:
OK. Done! I wrote a small method in `core/convert_DEIMOSsavfiles.py`. Please let me know if would be better to put it in another folder. This small script converted the .sav file in .fits files, which are now in `data/static_calibs/keck_deimos/`

jhennawi commented on pypeit/calibrations.py:
```diff
@@ -598,6 +598,9 @@ def get_slits(self):
         self._chk_set(['det', 'calib_ID', 'par'])
 
+        # Use slit-mask design if spectrograph is `keck_deimos`
+        maskdesign = True if self.spectrograph.spectrograph == 'keck_deimos' else False
```
Lines like if spectrograph == keck_deimos are not really allowed in our coding model. This is why we have parsets and spectrograph files. My suggestion is to implement the a maskdesign parameter into the EdgeTrace parset and set this in the keck_deimos spectrograph file.

deborape replied:
Added maskdesign parameter into the EdgeTrace parset

jhennawi commented on pypeit/edgetrace.py:
```diff
@@ -696,6 +702,9 @@ def auto_trace(self, bpm=None, debug=False, show_stages=False):
               traces* are based on the result of :func:`peak_refine`.
             - Synchronize the left and right traces into pairs that
               define slit apertures using :func:`sync`.
+            - Use :func:`maskdesign_matching` to add match the slit
```
Should this function maskdesign_matching be a method in spectrograph? It seems to me it should.

jhennawi commented on pypeit/edgetrace.py:
```diff
@@ -787,6 +799,12 @@ def auto_trace(self, bpm=None, debug=False, show_stages=False):
+        # the slit-mask design. If we include the piece of code in which `maskdesign_matching` recovers
+        # missing traces, we should move the following 2 lines before `self.sync()`
+        if maskdesign is True:
+            self.maskdesign_matching(offsets_range=[-50, 50], step=1, debug=False)
```
I would add the parset parameters now, as otherwise we are breaking our design rules about spectrograph specific hooks in code (see my point above about calibrations).

deborape replied:
Added to the parset parameters in `EdgeTracePar`

jhennawi commented on pypeit/edgetrace.py:
```diff
@@ -3970,326 +3988,235 @@ def fully_masked_traces(self, flag=None, exclude=None):
+        slit_id = self.spectrograph.slitmask.slitid
+
+        npoints = 250
+        nslits = self.spectrograph.slitmask.nslits
```
Why are you assuming the wavelengths are fixed here, I'm lost?

deborape replied:
This is hard-coded for the DEIMOS optical model. I moved this part to `mask_to_pixel_coordinates()`

jhennawi commented on pypeit/edgetrace.py:
```diff
@@ -3970,326 +3988,235 @@ def fully_masked_traces(self, flag=None, exclude=None):
+            msgs.error('Unable to read grating info')
+        if self.spectrograph.get_amapbmap(self.traceimg.files[0]) is None:
+            msgs.error('Unable to read amap and bmap')
+
```
Avoid using x and y to describe positions in PypeIt imags. The preferred terminology is always, (spec, spat), where spectral is the first dimension of PypeIt spectra, which is always spectral and spat is the second dimension of PypeIt oriented spectra which is always spatial. By sticking to this convention  you avoid ambiguity between flipped images in python, IDL, etc.

jhennawi commented on pypeit/edgetrace.py:
```diff
@@ -3970,326 +3988,235 @@ def fully_masked_traces(self, flag=None, exclude=None):
+        # top
+        xt_img, yt_img, ccd_t, xt_pix, yt_pix = self.spectrograph.mask_to_pixel_coordinates(x=xtop_mm, y=ytop_mm,
+                                                                                            wave=wave)
+
```
Does taking the median perform better than just evaluating the trace at the midpoint spectral position?

deborape replied:
I haven't really investigated that.

jhennawi commented on pypeit/edgetrace.py:
```diff
@@ -3970,326 +3988,235 @@ def fully_masked_traces(self, flag=None, exclude=None):
+                                                    xcen_mm[sortindx][i], model_xb[sortindx][i], model_xt[sortindx][i]))
+
+        reference_row = self.left_pca.reference_row if self.par['left_right_pca'] else self.pca.reference_row
+        xb_det = self.edge_fit[reference_row,self.is_left]
```
Please change all uses of x and y in variable names in this routine to spat and spec.

deborape replied:
Done!

jhennawi commented on pypeit/spectrographs/keck_deimos.py:
```diff
@@ -8,9 +8,12 @@
 import warnings
 
 from scipy import interpolate
+from scipy.io import readsav
```
See my previous comment about IDL save files.

jhennawi commented on pypeit/spectrographs/opticalmodel.py:
```diff
@@ -11,12 +14,12 @@ class ReflectionGrating:
     Doc this!
     """
```
Please add docs!

jhennawi reviewed (changes requested):
Great work @deborape. Minor comments.

deborape commented on pypeit/edgetrace.py:
```diff
@@ -3970,326 +3988,235 @@ def fully_masked_traces(self, flag=None, exclude=None):
+        ytop_mm = self.spectrograph.slitmask.top[:, 1]
+        ybot_mm = self.spectrograph.slitmask.bottom[:, 1]
+
+        slitindx = self.spectrograph.slitmask.slitindx
```
Removed

deborape commented on pypeit/edgetrace.py:
```diff
@@ -3970,326 +3988,235 @@ def fully_masked_traces(self, flag=None, exclude=None):
+
+        npoints = 250
+        nslits = self.spectrograph.slitmask.nslits
+        wave = np.arange(npoints) * 24. + 4000.
```
Done

deborape commented on pypeit/calibrations.py:
```diff
@@ -598,6 +598,9 @@ def get_slits(self):
         self._chk_set(['det', 'calib_ID', 'par'])
 
+        # Use slit-mask design if spectrograph is `keck_deimos`
+        maskdesign = True if self.spectrograph.spectrograph == 'keck_deimos' else False
```
Done!

deborape commented on pypeit/edgetrace.py:
```diff
@@ -3970,326 +3988,235 @@ def fully_masked_traces(self, flag=None, exclude=None):
+        # Find if there are missing traces.
+        # Need exactly one occurrence of each index in "need"
+        buffer = 20.
+        need = ((top_edge_pred > buffer) & (bot_edge_pred < 2047. - buffer)) & ((model_xb != -1) | (model_xt != -1))
```
Done.

deborape commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,226 @@
+    coeff=numpy.append(coeff, 0)
+    yfit = slit_coeff_eval(x_model[ind], x_model[ind]*0., coeff)
+
+    sigrej = 3
```
Added to Parset.

deborape commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,226 @@
+
+        sdev[j] = (offs**2).sum()          # big if match is bad
+
+    best_sdev = int(numpy.mean(numpy.where(sdev==sdev.min())[0]))      # average ind
```
Right!

deborape replied:
Although, I feel more comfortable to keep the mean, just in case there is a remote case where more than one lag gives the exact same minimum.

deborape commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,226 @@
+
+    # check for duplicate indices
+    bad = numpy.ones(ind.shape[0], dtype=int)
+    bad[numpy.unique(ind, return_index=True)[1]] = 0
```
It was hard to understand for me too! 
But, I now understand that, if there is a duplicate (which I feel there should not be, at least not often), the code is not, actually, assuming that the first one is Ok and discard the others. Instead, in the `for` loop below the code actually selects only the duplicates that have the smallest residuals to the fit. I put some more inline comments here to better understand it. 

It seems that there could be duplicates if an edge is found where there is not one. Those duplicates are then trimmed, but this trimming needs to be recorded in the `maskdef_id`. I am adding a `-99` value for those. I am assuming that this will not happen very often, but I'll do more tests.

deborape commented:
Alright! @profxj, @jhennawi, @kbwestfall,

I am done with addressing all the comments. Could you, please, two of you (or all of you) review them.

-- Added Parset for the slit-mask design matching.
-- Added a QA printed on the screen that lists the slit IDs in that specific detector
-- converted .sav files into .fits files 
-- docs added everywhere

Dev Suite `./pypeit_test develop -i keck_deimos` passed
`--- PYPEIT DEVELOPMENT SUITE PASSED 7/7 TESTS (Masters ignored) ---`

I am now checking the log files of the Dev Suite runs to check that everything looks reasonable.

profxj reviewed (approved):
Have read through the new changes and have no complaints,
only thanks. :)

I approve.

kbwestfall commented on pypeit/core/convert_DEIMOSsavfiles.py:
```diff
@@ -0,0 +1,33 @@
+from scipy.io import readsav
```
I doubt we'll need this function again, but we might want to move it to `pypeit.io`.  I would leave it for now.

kbwestfall reviewed (changes requested):
Thanks, Debora!  I have some minor comments.  I also get these test failures:
```
=========================== short test summary info ============================
FAILED test_opticalmodel.py::test_deimos_mask_coordinates - ValueError: The t...
FAILED test_opticalmodel.py::test_deimos_ccd_slits - AssertionError: Should f...
================== 2 failed, 245 passed in 1120.65s (0:18:40) ==================
```

It seems that shape of the arrays returned by `mask_to_pixel_coordinates` are not as expected, and I wasn't sure how you would want to handle this.

deborape commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,331 @@
+            the mask design and the traced slit positions.
+
+    Returns:
+        xlag (:obj:`float`):
```
Thanks!

deborape commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,331 @@
+    yref_xdet = numpy.tile(yref, x_det.size)
+    yref_x_model = numpy.tile(yref, x_model.size)
+
+    _shape = (4096, 2048) if shape is None else shape
```
Done! Thanks!

deborape commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,331 @@
+        if trimmed is not None and x in trimmed:
+            plt.plot(edgetrace.T[x,:], numpy.arange(edgetrace.T[x,:].size), color='orange', lw=0.5, zorder=0)
+        else:
+            plt.plot(edgetrace.T[x, :], numpy.arange(edgetrace.T[x, :].size), color='k', lw=0.5, zorder=0)
```
Yes, I can!

deborape commented on pypeit/spectrographs/opticalmodel.py:
```diff
@@ -54,7 +55,58 @@ def _get_grating_transform(self):
                             [ cosx*sinr, -sint*cosr+cost*sinx*sinr,  cost*cosr+sint*sinx*sinr]])
 
-    def reflect(self, r, wave=None, order=1):
+    # def reflect(self, r, wave=None, order=1):
```
Ok!

deborape commented on pypeit/spectrographs/opticalmodel.py:
```diff
@@ -532,16 +718,62 @@ def mask_to_imaging_coordinates(self, x, y, wave, order):
+        # return self.grating_output_vectors_to_ics_coo(r, sign=1 - 2 * (x < 0))
+        return self.post_grating_vectors_to_ics_coo(r, bmap, nslits, npoints)
+
+    # def mask_to_imaging_coordinates(self, x, y, wave, order):
```
OK!

deborape commented on pypeit/edgetrace.py:
```diff
@@ -3971,326 +3993,293 @@ def fully_masked_traces(self, flag=None, exclude=None):
+            whgood = np.where(tedge_img[i, :] > -1e4)[0]
+            npt_img = int(whgood.shape[0] / 2.)
+            whgood = whgood[:npt_img] if ccdnum <= 4 else whgood[npt_img:]
+            if (omodel_bspat[i] == -1) & (omodel_tspat[i] >= 0):
```
Done!

deborape commented on pypeit/edgetrace.py:
```diff
@@ -3971,326 +3993,293 @@ def fully_masked_traces(self, flag=None, exclude=None):
+            # If a left (or right) trace is outside of the detector, the corresponding right (or left) trace
+            # is determined using the pixel position from the image plane.
+            whgood = np.where(tedge_img[i, :] > -1e4)[0]
+            npt_img = int(whgood.shape[0] / 2.)
```
Yes!

deborape commented on pypeit/edgetrace.py:
```diff
@@ -4341,11 +4328,391 @@ def _fill_objects_table(self, register):
 
+
+
+        # def mask_refine(self, design_file=None, allow_resync=False, debug=False):
```
OK!

deborape commented on pypeit/edgetrace.py:
```diff
@@ -3970,326 +3988,235 @@ def fully_masked_traces(self, flag=None, exclude=None):
+
+
+
+        if debug is True:
```
Done!

kbwestfall reviewed (approved):
Thanks for the test fixes!

deborape commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,331 @@
+            # as assumes don't match 2 x_det's to the same x_model
+            offs = numpy.zeros(x_det.size)
+            for xelement in range(x_det.size):
+                offs[xelement] = numpy.min(numpy.abs(x_det_lag[xelement]-x_model_trim))
```
Thanks!

deborape commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,331 @@
+    # for each traced edge (`x_det`) determine the value of x_model that gives the smallest offset
+    ind = numpy.zeros(x_det.size, dtype=int)
+    for i in range(ind.size):
+        ind[i] = numpy.argmin(numpy.abs(x_det[i] - x_model_new))
```
Done here too. Thanks!

deborape commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,331 @@
+    out = numpy.abs(res) > cut
+
+    # check for duplicate indices
+    bad = numpy.ones(ind.size, dtype=int)
```
Done!

jhennawi commented on pypeit/core/slitdesign_matching.py:
```diff
@@ -0,0 +1,331 @@
+"""
+# from IPython import embed
+
+import numpy
```
We use import numpy as np everywhere in pypeit, so I would not suggest we deviate from that convention.

deborape commented:
Alright! @profxj, @kbwestfall, @jhennawi 

The unit tests `test_maskdesign_matching.py ` and `test_opticalmodel.py ` passed.

Dev Suite `./pypeit_test develop -i keck_deimos` passed.
`--- PYPEIT DEVELOPMENT SUITE PASSED 7/7 TESTS (Masters ignored) ---`

I checked all the results from the Dev Suite, and compared them with what I get from the IDL pipeline. This allowed me to spot some needed tweaks. I fixed those. Now the code should be ready!

If any of you wish to double check, I will not mind it at all. Otherwise I think we can merge.

jhennawi reviewed (changes requested):
Hi Gang, sorry to be the odd man out here, but these changes to edgetrace look significant that I think you need to run the entire dev suite. Please run and post those tests and I will approve. 

Thanks!
Joe

deborape commented:
> Hi Gang, sorry to be the odd man out here, but these changes to edgetrace look significant that I think you need to run the entire dev suite. Please run and post those tests and I will approve.
> 
> Thanks!
> Joe

@jhennawi, @profxj, @kbwestfall sorry for taking so long, but I finally run the entire Dev Suite. Below is the result. I got 2 failed tests but none of them involves the main reduction (and `EdgeTrace`). 
The problem with `pypeit_sensfunc` was due to the fact that I did not have the atmosphere telluric models saved on my computer.
The problem with `pypeit_coadd_2dspec` was due to a VLT naming issue that I have with my computer (which does not like file names with commas). I fixed it for most of the `.pypeit` files, but I missed this one `.coadd2d` file.

`--- PYPEIT DEVELOPMENT SUITE FAILED 2/84 TESTS (Masters ignored) ---`
`Failed tests:`
`    gemini_gnirs/32_SB_SXD pypeit_sensfunc`
`    vlt_xshooter/VIS_manual pypeit_coadd_2dspec`
`Skipped tests:`
`    gemini_gnirs/32_SB_SXD pypeit_flux_setup`
`    gemini_gnirs/32_SB_SXD pypeit_flux`
`    gemini_gnirs/32_SB_SXD pypeit_coadd_1dspec`
`    gemini_gnirs/32_SB_SXD pypeit_coadd_2dspec`
`    gemini_gnirs/32_SB_SXD pypeit_tellfit`

profxj reviewed (approved):
Ready to merge when tests pass.

deborape commented:
New run of the Dev Suite. Unfortunately still one failed test.

```
--- PYPEIT DEVELOPMENT SUITE FAILED 1/89 TESTS (Masters ignored) ---
Failed tests:
    lbt_luci/LUCI-I pypeit (without masters)
Skipped tests:
Testing Started at 2020-10-10T12:09:37.138900
Testing Completed at 2020-10-11T11:35:19.891062
Total Time: 23:25:42.752162
```

profxj commented:
Passing on my laptop:

(base) wolverine> ./pypeit_test -t 2 -i lbt_luci develop
Running tests on the following instruments:
    lbt_luci

Reducing data from lbt_luci for the following setups:
    LUCI-I
    LUCI-II

Running tests in 2 parallel processes
 0 passed/ 0 failed/ 0 skipped STARTED lbt_luci/LUCI-I pypeit (without masters)
 0 passed/ 0 failed/ 0 skipped STARTED lbt_luci/LUCI-II pypeit (without masters)
 1 passed/ 0 failed/ 0 skipped PASSED  lbt_luci/LUCI-II pypeit (without masters)
 2 passed/ 0 failed/ 0 skipped PASSED  lbt_luci/LUCI-I pypeit (without masters)

--- PYPEIT DEVELOPMENT SUITE PASSED 2/2 TESTS (Masters ignored) ---

profxj commented:
merging
